### DISCUSSION TITLE
I would be seriously grateful for a Swift <--> C++ integration friend

### DISCUSSION THREAD

**Discussion by dougdew64** [ml-explore/mlx#112] [General]
I am attempting to implement a swift equivalent of the c++ tutorial application.

I am getting my butt kicked in Xcode trying to figure out how to correctly structure and configure my projects in order to make Xcode happy. After several hours, I found a structure and configuration which Xcode doesn't complain about. 

Now I'm encountering a surprising compilation failure on my first line of tutorial code. The swift compiler complains that there's no mlx.core.array (mlx::core::array) available. I've attempted this compilation from the command line and experienced the same failure.

Interestingly, the Xcode intellisense seems to be aware of mlx.core.array. 

![image](https://github.com/ml-explore/mlx/assets/98898390/d497b266-2062-4b86-a8e2-9a2fc28ce8a5)

**Comment by RahulBhalley** [comment]
This would be really interesting as S4TF was killed unexpectedly. 

- Author of _Deep Learning with Swift for TensorFlow_ book here

**Comment by dougdew64** [comment]
@RahulBhalley you might find this discussion interesting: https://github.com/ml-explore/mlx/issues/15

**Comment by RahulBhalley** [comment]
Hey! Any success in running MLX with C++ project in Xcode?

**Comment by dougdew64** [comment]
@RahulBhalley I used MLX, C++ and Xcode a while back. https://github.com/ml-explore/mlx/discussions/86

**Comment by dougdew64** [comment]
I've been pretty much stuck since then. I had been intending to add higher-level C++ and eventually implement an all-C++ version of the LLM example. However, a few weeks ago @awni and @davidkoski suggested that the Swift API would be available soon. So, I paused my effort to await the Swift API. When the Swift API is available, then I'll resume my effort while using the Swift API as a guide for my C++ effort. For example, I want to see how details such as array slicing are implemented in Swift before attempting the same in C++.

**Reply by davidkoski** [reply]
Depending on what you mean, you can take a look at the implementation of the python binding:

- https://github.com/ml-explore/mlx/blob/main/python/src/indexing.cpp#L378

`mlx_get_item_slice` is `array[1:10]` in python.  This in turn calls slice() in the C++ back end:

- https://github.com/ml-explore/mlx/blob/main/mlx/ops.cpp#L439 

The swift implementation could call the same code in the C++ back end for its implementation of `array[1 ..< 10]`

That doesn't really answer how to call that from swift (larger issue) but it would let you write a C++ version of python code that was slicing (which is what I understood the question to be asking).

**Reply by dougdew64** [reply]
Thanks @davidkoski.

I had been thinking along the lines which you described, but wanted to first see how you solve the same problem(s) in the Swift binding. 

Are you implementing in Swift the same kind of logic which is contained in indexing.cpp?

**Reply by davidkoski** [reply]
Some of it -- the python API allows you to mix all of the different indexing operations at once.  The swift version is typed and you would just compose them or use different operators.  If you look at `mlx_get_item_nd()` that is what the python side does internally.

In practice, e.g. porting some python code for testing, this has worked out ok.  It doesn't preclude doing `subscript(Any...)` in the future.  I think in C++ it might be a little easier with `std::variant`.

**Reply by dougdew64** [reply]
Thanks @davidkoski
